File: tracks/typescript/src/inventory-item.ts
export class InventoryItem {
  public constructor(
    public readonly sku: string,
    public readonly name: string,
    public quantity: number,
    public readonly reorderLevel: number,
  ) {}

  public get isLowStock(): boolean {
    return this.quantity <= this.reorderLevel;
  }

  public reserve(quantity: number): void {
    if (quantity > this.quantity) {
      throw new Error("Insufficient stock");
    }

    this.quantity -= quantity;
  }
}

File: tracks/dotnet/Inventory.Domain/InventoryItem.cs
namespace Inventory.Domain;

public sealed class InventoryItem(
    string sku,
    string name,
    int quantity,
    int reorderLevel)
{
    public string Sku { get; } = sku;
    public string Name { get; } = name;
    public int Quantity { get; private set; } = quantity;
    public int ReorderLevel { get; } = reorderLevel;
    public bool IsLowStock => Quantity <= ReorderLevel;

    public void Reserve(int quantity)
    {
        if (quantity > Quantity)
        {
            throw new InvalidOperationException("Insufficient stock");
        }

        Quantity -= quantity;
    }
}
